internal/stores/hostdb.go

```diff
@@ -131,26 +133,50 @@ func (db *SQLStore) Host(hostKey consensus.PublicKey) (hostdb.Host, error) {
 	return h.convert(), tx.Error
 }
 
-// Hosts returns up to max hosts that have not been interacted with since
-// the specified time.
-func (db *SQLStore) Hosts(notSince time.Time, max int) ([]hostdb.Host, error) {
-	// Filter all hosts for the ones that have not been updated since a
-	// given time.
-	var fullHosts []dbHost
-	err := db.db.Table("hosts").
-		Joins("LEFT JOIN host_interactions ON host_interactions.db_host_id = hosts.ID").
-		Group("Public_Key").
-		Having("IFNULL(MAX(Timestamp), 0) < ?", notSince.UTC()). // use UTC since we stored timestamps in UTC
-		Limit(max).
+// Hosts returns hosts at given offset and limit.
+func (db *SQLStore) Hosts(offset, limit int) ([]hostdb.Host, error) {
+	if offset < 0 {
+		return nil, ErrNegativeOffset
+	}
+
+	var dbHosts []dbHost
+	err := db.db.
+		Model(&dbHost{}).
+		Offset(offset).
+		Limit(limit).
 		Preload("Interactions").
 		Preload("Announcements").
-		Find(&fullHosts).
+		Find(&dbHosts).
 		Error
 	if err != nil {
 		return nil, err
 	}
+
 	var hosts []hostdb.Host
-	for _, fh := range fullHosts {
+	for _, fh := range dbHosts {
+		hosts = append(hosts, fh.convert())
+	}
+	return hosts, err
+}
+
+// RandomHosts returns up to given limit of hosts that are randomly selected
+// from the host database.
+func (db *SQLStore) RandomHosts(limit int) ([]hostdb.Host, error) {
```

Comment:
I don't think we should bother with random host selection in the bus before we have protections against spammers. 
Imo we are better off if we allow for easy blacklisting of hostnames first and ship renterd with some known spammers on the blacklist. That will significantly reduce the number of hosts we return for scanning already.

Then the scanning code should be able to happily chew through the remaining hosts using pagination and prune the once which are offline for too long.

Finally the contract formation code will only need to fetch hosts that are neither blacklisted nor offline which should only be hundreds right now. Even thousands could be fetched at once without issues.

Reply:
A huge downside of the random host approach in general is that you pick from small random batches. So we won't be able to use a weighted binary tree like we do in siad anymore to probabilistically prefer hosts with a better score when picking them.